internal/unionstore/art/art_node.go

```diff
@@ -12,47 +12,660 @@
+		curr := n256.present[presentOffset] & mask
+		zeros := bits.TrailingZeros64(curr)
+		if zeros < n48m {
+			return presentOffset*n48m + zeros
+		}
+	}
+	return node256cap
+}
+
+func (n256 *node256) prevPresentIdx(start int) int {
+	for presentOffset := start >> n48s; presentOffset >= 0; presentOffset-- {
+		offset := start % n48m
+		start = n48m - 1
+		mask := uint64(1)<<(offset+1) - 1 // e.g. offset=3 => 0b000...0001111
+		curr := n256.present[presentOffset] & mask
+		zeros := bits.LeadingZeros64(curr)
+		if zeros < n48m {
+			return presentOffset*n48m + n48m - (zeros + 1)
+		}
+	}
+	return inplaceIndex
+}
+
+// key methods
+
+func (k artKey) charAt(pos int) byte {
+	if pos < 0 || pos >= len(k) {
+		return 0
+	}
+	return k[pos]
+}
+
+func (k artKey) valid(pos int) bool {
+	return pos < len(k)
+}
+
 // GetKey gets the full key of the leaf
 func (l *artLeaf) GetKey() []byte {
-	panic("unimplemented")
+	base := unsafe.Add(unsafe.Pointer(l), leafSize)
+	return unsafe.Slice((*byte)(base), int(l.klen))
+}
+
+// getKeyDepth gets the partial key start from depth of the artLeaf
+func (l *artLeaf) getKeyDepth(depth uint32) []byte {
+	base := unsafe.Add(unsafe.Pointer(l), leafSize+int(depth))
+	return unsafe.Slice((*byte)(base), int(l.klen)-int(depth))
 }
 
 // GetKeyFlags gets the flags of the leaf
 func (l *artLeaf) GetKeyFlags() kv.KeyFlags {
 	panic("unimplemented")
 }
+
+func (l *artLeaf) match(depth uint32, key artKey) bool {
+	return bytes.Equal(l.getKeyDepth(depth), key[depth:])
+}
+
+func (l *artLeaf) setKeyFlags(flags kv.KeyFlags) arena.MemdbArenaAddr {
+	l.flags = uint16(flags) & flagMask
+	return l.vAddr
+}
+
+func (l *artLeaf) getKeyFlags() kv.KeyFlags {
+	return kv.KeyFlags(l.flags & flagMask)
+}
+
+const (
+	deleteFlag uint16 = 1 << 15
+	flagMask          = ^deleteFlag
+)
+
+// markDelete marks the artLeaf as deleted
+//
+//nolint:unused
+func (l *artLeaf) markDelete() {
+	l.flags = deleteFlag
+}
+
+//nolint:unused
+func (l *artLeaf) unmarkDelete() {
+	l.flags &= flagMask
+}
+
+func (l *artLeaf) isDeleted() bool {
+	return l.flags&deleteFlag != 0
+}
+
+// node methods
+func (n *nodeBase) setPrefix(key artKey, prefixLen uint32) {
+	n.prefixLen = prefixLen
+	copy(n.prefix[:], key[:min(prefixLen, maxPrefixLen)])
+}
+
+func (n *nodeBase) match(key artKey, depth uint32) uint32 {
+	idx := uint32(0)
+
+	limit := min(min(n.prefixLen, maxPrefixLen), uint32(len(key))-depth)
+	for ; idx < limit; idx++ {
+		if n.prefix[idx] != key[idx+depth] {
+			return idx
+		}
+	}
+
+	return idx
+}
+
+func (an *artNode) node4(a *artAllocator) *node4 {
+	return a.getNode4(an.addr)
+}
+
+func (an *artNode) node16(a *artAllocator) *node16 {
+	return a.getNode16(an.addr)
+}
+
+func (an *artNode) node48(a *artAllocator) *node48 {
+	return a.getNode48(an.addr)
+}
+
+func (an *artNode) node256(a *artAllocator) *node256 {
+	return a.getNode256(an.addr)
+}
+
+func (an *artNode) matchDeep(a *artAllocator, key artKey, depth uint32) uint32 /* mismatch index*/ {
```

Comment:
May need some comments for it, particularly to distinguish its functionality from that of the `match`